src/jit/hwintrinsicxarch.cpp

```diff
@@ -418,7 +418,7 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
     HWIntrinsicCategory category = categoryOfHWIntrinsic(intrinsic);
     HWIntrinsicFlag     flags    = flagsOfHWIntrinsic(intrinsic);
     int                 numArgs  = sig->numArgs;
-    var_types           retType  = JITtype2varType(sig->retType);
+    var_types           retType  = JITtype2varTypeExact(sig->retType);
```

Comment:
As far as I can tell `retType` is a scalar type and the type of the intrinsic node that will be created. I can understand the need for uint/ulong as a SIMD base type but it's not clear if there's any need to create nodes having these types. I don't think the JIT ever created such nodes until now so this might cause problems.

Reply:
> I don't think the JIT ever created such nodes until now so this might cause problems.

Actually, GT_CAST (GenTreeCast) nodes are using `uint`/`ulong` all the time. This usage is central to table driven HW intrinsic import and code gen. We have to be able to use `exact` `retType` type for instruction emitting. Scenario where we use `retType` for emmision is a base one, use of `argType` type is an alternative scenario used in several cases.

Reply:
> Actually, GT_CAST (GenTreeCast) nodes are using uint/ulong all the time

They use uint/ulong in a very specific manner. In particular, they don't use these type as `gtType`.

> This usage is central to table driven HW intrinsic import and code gen. We have to be able to use exact retType type for instruction emitting. Scenario where we use retType for emmision is a base one, use of argType type is an alternative scenario used in several cases.

Examples?

Reply:
Every table driven hw intrinsic which is marked with category `HW_Category_SimpleSIMD` without flag `HW_Flag_BaseTypeFromArg` uses `retType` as `baseType` (except where baseType has to be determined by `getBaseTypeOfSIMDType`), on the other hand every  `HW_Category_SimpleSIMD` with flag `HW_Flag_BaseTypeFromArg` uses `argType` as `baseType`.

Reply:
Hmm...
* apparently there's a ton of them
* but many of them return vectors, not scalars, so `retype` would be TYP_STRUCT
* and anyway it's strange to depend on the return type, with a few exceptions it's the parameter type(s) that should decide what operation is performed

Reply:
> but many of them return vectors, not scalars, so retype would be TYP_STRUCT

For all such cases we use not always directly `getBaseTypeAndSizeOfSIMDType` or `getBaseTypeOfSIMDType`.

The above selection criteria were just an example. Most scalar intrinsics marked with `HW_Category_SIMDScalar` will need that path as well.

Reply:
> There is handling to get the TYP_SIMD rather than TYP_STRUCT...

All that stuff happens after this line.